### PR TITLE
Update Linux support to 0.0.13

### PR DESCRIPTION
**Pull request by sarnex**
Code heavily based on @Francesco149's original PR, with line numbers updated.

Signed-off-by: Nick Sarnie commendsarnex@gmail.com

His original instructions are below, with slight modification:
**Make sure you have at least node v5**

```


cd
mkdir discord
cd discord

# Install discord in /opt/
wget https://discordapp.com/api/download/canary?platform=linux&format=tar.gz
tar -xzvf discord-canary-0.0.3.tar.gz
sudo cp -r DiscordCanary /opt/
sudo ln -s /opt/DiscordCanary/DiscordCanary /usr/bin/discord

# Install BetterDiscord
git clone https://github.com/sarnex/BetterDiscordApp.git
mv BetterDiscordApp BetterDiscord
cp BetterDiscord/Installers/Node/* .
cp BetterDiscord/splice .
npm install asar wrench
sudo ./install.sh
```

### CODE REVIEW

File: NodeInstaller/index.js [hunk 1]
```diff
@@ -1,248 +1,255 @@
...
+            _importSplice = 67;
+            _functionCallSplice = 446;
+            _functionSplice = 547;
+            _discordPath = "/Applications/Discord.app/Contents/Resources"; // Defaults to Applications directory
+        } else if (_os == "linux") {
+			_importSplice = 63;
+			_functionCallSplice = 507;
+			_functionSplice = 580;
+			_discordPath = "/opt/DiscordCanary/resources";
+			// not sure where the .deb install, and most people are going to
```
Comment (DaFri-Nochiterov) [inline NodeInstaller/index.js]: AFAIK `*.deb` file installs in `/usr/share/discord-canary`

Comment (sarnex) [inline NodeInstaller/index.js]: The original required the user to install in /opt, and not everyone is on Debian or Ubuntu, so I'm inclined to leave that for now.

**Comment by DaFri-Nochiterov** [comment]
```bash
cd

# Move dir if exists
oldTime=$(date +"%H-%M-%S.%d-%m-%y")
mv discord "discord.${oldTime}.old"

# And.. Create new!
mkdir discord
cd discord

# Install discord in /opt/
wget "https://discordapp.com/api/download/canary?platform=linux&format=tar.gz" -O discord-canary-0.0.3.tar.gz
tar -xzvf discord-canary-0.0.3.tar.gz
sudo cp -r DiscordCanary /opt/
sudo ln -s /opt/DiscordCanary/DiscordCanary /usr/bin/discord

# Install BetterDiscord
git clone https://github.com/sarnex/BetterDiscordApp.git
mv BetterDiscordApp BetterDiscord
cp BetterDiscord/NodeInstaller/* .
cp BetterDiscord/splice .
npm install asar wrench
sudo ./install.sh
```

**Comment by Francesco149** [comment]
ah thanks for updating my pull request, was too busy to work on it

**Comment by mkody** [comment]
Works fine for me!

I needed to change the `_discordPath` in the `NodeInstaller/index.js` file since the [AUR package for Discord-Canary](https://aur.archlinux.org/packages/discord-canary/) uses `/opt/discord-canary/`. 
But it looks like this path is not consistent between distributions. But for Arch, someone might just add a patch during installation if they make a AUR package, if needed :+1:

**Comment by sarnex** [comment]
Still works with no change on version 0.4

**Comment by DaFri-Nochiterov** [comment]
Not works on newest 0.0.6 version :(

**Comment by sarnex** [comment]
@DaFri-Nochiterov should be fixed now

**Comment by DaFri-Nochiterov** [comment]
@sarnex, yes, this works. Thanks.

**Comment by DaFri-Nochiterov** [comment]
Again. Not works on 0.0.7 version.

Now with `Exception`:
```javascript
Uncaught Exception:
SyntaxError: Unexpected token ;
    at exports.runInThisContext (vm.js:53:16)
    at Module._compile (module.js:387:25)
    at Object.Module._extensions..js (module.js:422:10)
    at Module.load (module.js:357:32)
    at Function.Module._load (module.js:314:12)
    at Object.<anonymous> (/opt/DiscordCanary/resources/electron.asar/browser/init.js:184:8)
    at Module._compile (module.js:413:34)
    at Object.Module._extensions..js (module.js:422:10)
    at Module.load (module.js:357:32)
    at Function.Module._load (module.js:314:12)
```

**Comment by sarnex** [comment]
@DaFri-Nochiterov fixed again, hopefully they leave index.js alone. If not, if it happens again, I'll look for a better solution than hardcoding line numbers, as is done in the official installer.

**Comment by mkody** [comment]
@sarnex You should definitely look for a better solution, because I'm pretty sure the index.js will change a lot in the near future.

**Comment by sarnex** [comment]
I've rewritten the linux installer to not use hardcoded line numbers. It should not break now unless they remove the two lines I'm searching for, which I don't expect.

**Comment by mkody** [comment]
And it works fine on just-released `0.0.8`! 👍

**Comment by DaFri-Nochiterov** [comment]
Emoticons no longer works in this version

**Comment by sarnex** [comment]
No idea why, installer doesn't seem to have been modified. Someone else should investigate.

**Comment by mkody** [comment]
Huh? It works for me. Running from commit 6be4da2197ea57025f653a07903fa01764110c70

**Comment by sarnex** [comment]
Emotes work now, it was a BetterDIscord bug unrelated to these changes.

**Comment by simoniz0r** [comment]
I have tried installing this multiple times, but I keep getting this error after running `sudo ./install.sh`

`simonizor@T00L-Xubuntu:~/Downloads/discord$ sudo ./install.sh 
/home/simonizor/Downloads/discord/index.js:128
 fs.exists(process.env.HOME + _prefsPath + emotes_twitch_global, (exists) => {
                                                                           ^
SyntaxError: Unexpected token >
    at Module._compile (module.js:439:25)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Function.Module.runMain (module.js:497:10)
    at startup (node.js:119:16)
    at node.js:902:3
simonizor@T00L-Xubuntu:~/Downloads/discord$
`

Any help would be appreciated.

**Comment by chowder908** [comment]
I'm getting the same error as simoniz0r.

**Comment by sarnex** [comment]
@simoniz0r @chowder908 

I've rebased the pull request based on the lastest BetterDiscord, and I tried the install from scratch and it worked. I've also updated the original instructions to fit the current directory structure. 

Let me know if there's still an issue.

**Comment by chowder908** [comment]
sudo: ./install.sh: command not found

Now telling me this.

**Comment by chowder908** [comment]
Is there anyway to just manually install it like copying the files to the places they have to be?

**Comment by chowder908** [comment]
@sarnex  sudo: ./install.sh: command not found
Now telling me this.
Is there anyway to just manually install it like copying the files to the places they have to be?

**Comment by sarnex** [comment]
@chowder908 you have to chmod +x install.sh

**Comment by chowder908** [comment]
@sarnex 
```
sudo chmod +x ./install.sh
chmod: cannot access ‘./install.sh’: No such file or directory

```

**Comment by sarnex** [comment]
@chowder908 did you follow the instructions above? 
````
git clone https://github.com/sarnex/BetterDiscordApp.git
mv BetterDiscordApp BetterDiscord
cp BetterDiscord/Installers/Node/* .
cp BetterDiscord/splice .
npm install asar wrench
sudo ./install.sh
````

**Comment by chowder908** [comment]
@sarnex line by line

**Comment by simoniz0r** [comment]
```
simonizor@T00L-Xubuntu:~/Downloads/discord$ sudo ./install.sh 

/home/simonizor/Downloads/discord/index.js:126
 fs.exists(process.env.HOME + _prefsPath + emotes_twitch_global, (exists) => {
                                                                           ^
SyntaxError: Unexpected token >
    at Module._compile (module.js:439:25)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Function.Module.runMain (module.js:497:10)
    at startup (node.js:119:16)
    at node.js:902:3
```
I'm getting pretty much the same error after deleting everything and cloning your PR after these changes.

**Comment by chowder908** [comment]
Yeah @simoniz0r i redid it and same error.

**Comment by sarnex** [comment]
@simoniz0r Can you send me the output of node -v? I think it's some weird library or node version issue.

**Comment by simoniz0r** [comment]
```
simonizor@T00L-Xubuntu:~/Downloads/discord$ node -v
v0.10.25
```

**Comment by chowder908** [comment]
@sarnex I have v0.10.29

**Comment by sarnex** [comment]
That version is from around 2014. See if you can update it. I'm on 5.10.1

**Comment by simoniz0r** [comment]
@sarnex Thank you, that worked!

@chowder908 In case you need help updating node, this is how I did it:
```
sudo npm cache clean -f
sudo npm install -g n
sudo n stable

sudo ln -sf /usr/local/n/versions/node/<VERSION>/bin/node /usr/bin/node 
```

**Comment by chowder908** [comment]
@sarnex @simoniz0r i'm getting sudo: npm: command not found

**Comment by sarnex** [comment]
@simoniz0r cool, glad it worked. I'll add it to the original post

File: Installers/Node/index.js [hunk 2]
```diff
@@ -99,8 +101,18 @@ function install() {
                     console.log("Injecting index.js");
 
                     var data = fs.readFileSync(_discordPath + _index).toString().split("\n");
-                    data.splice(_importSplice, 0, 'var _betterDiscord = require(\'betterdiscord\');\n');
-                    data.splice(_functionCallSplice, 0, splice);
+                    
+                    if(_os == "linux") {
```
Comment (Tsunamical) [inline Installers/Node/index.js]: darwin needs this now too

Comment (Tsunamical) [inline Installers/Node/index.js]: Actually, we can just remove the if clause here, #336 has it by default. Should be functional across the board.

**Comment by carlmiller307** [comment]
I installed this on my copy of DiscordCanary, which runs on Ubuntu 16.04 x64, and the load process hangs on FrankerFaceZ emote data.

**Comment by mkody** [comment]
@carlmiller307 Your problem looks unrelated to this pull request. Here we're talking about the installation script. If it says that FrankerFaceZ loads, that mean it's installed (but has a bug).
You should try to [open a new issue](https://github.com/Jiiks/BetterDiscordApp/issues/new) for that.

**Comment by GhettoGirl** [comment]
I'm getting this error in the Discord app 0.0.8


>(program):1 Uncaught TypeError: Cannot read property 'appendChild' of undefinedelectron.ipcRenderer.on @ /opt/discord-canary/resources/electron.asar/renderer/init.js:44
>(program):1 Uncaught TypeError: Cannot set property 'innerHTML' of nullelectron.ipcRenderer.on @ /opt/discord-canary/resources/electron.asar/renderer/init.js:44
>(program):1 Uncaught TypeError: Cannot set property 'value' of nullelectron.ipcRenderer.on @ /opt/discord-canary/resources/electron.asar/renderer/init.js:44
>(program):1 Uncaught TypeError: Cannot set property 'max' of nullelectron.ipcRenderer.on @ /opt/discord-canary/resources/electron.asar/renderer/init.js:44


Node v6.4.0
Arch Linux

I applied this pull request on top of **master** (48d4a58102dcdbc1ce888b322df2f7aa834eb695)

**Comment by mkody** [comment]
@GhettoGirl Hmm... Works fine for me, I'm using the `discord-canary` AUR package.
```
yaourt -S discord-canary
# I've started Discord at least once and then quit before installing BetterDiscord
mkdir BDA && cd BDA
git clone https://github.com/Jiiks/BetterDiscordApp.git && cd BetterDiscordApp
git remote add sarnex https://github.com/sarnex/BetterDiscordApp.git
git pull sarnex master
cd .. && mv BetterDiscordApp BetterDiscord
cp BetterDiscord/Installers/Node/* BetterDiscord/splice .
npm install asar wrench
sudo ./install.sh
```

**Comment by GhettoGirl** [comment]
@mkody I restarted Discord several times, now it works. It may be related to this upstream bug https://github.com/crmarsh/discord-linux-bugs/issues/81. But this error is new, I'm gonna report this in the upstream issue too.

**Comment by jamesguessis** [comment]
do plugins and themes work with this?

**Comment by mkody** [comment]
@jamesguessis Of course, yes. What's the point if they don't work? :smile:

**Comment by jamesguessis** [comment]
where would the respective folders be located then?

**Comment by mkody** [comment]
Inside `~/.config/BetterDiscord`

**Comment by GhettoGirl** [comment]
Discord Canary 0.0.9 released and BetterDiscord is no longer compatible with it. Says `_betterDiscord` is undefined. Just want to let you know.

I'm staying at 0.0.8 for a while.

**Comment by sarnex** [comment]
Sorry for the delay, I've been busy lately. I've updated the PR for 0.0.9. 

It works for me, please let me know if it works.

Thanks.

**Comment by GhettoGirl** [comment]
It works great. Thanks for the quick update. :+1:

**Comment by mkody** [comment]
Can confirm it works. :+1:

**Comment by simoniz0r** [comment]
Thanks for the update.  I was having some issues with the install script not creating a directory properly before and also the crash above with 0.0.9.  Both of those issues were resolved with this update.  The install went as expected, and no crash on loading 0.0.9.

**Comment by sarnex** [comment]
You Rock!
Cancel That
I'm the greatest

**Comment by sarnex** [comment]
Works with 0.0.10 without changes.

**Comment by mkody** [comment]
Works fine with 0.0.11 too.

**Comment by sarnex** [comment]
@mkody thanks for testing

**Comment by mkody** [comment]
Aaaaand, works fine with 0.0.12!

**Comment by sarnex** [comment]
Tubular

**Comment by magiruuvelvet** [comment]
Still works with 0.0.13

**Comment by TheDoctorsLife** [comment]
Any possibility of getting this merged into master? It seems this is pretty solid and plenty of people are reporting it still working with the latest version of Discord Linux.

**Comment by sarnex** [comment]
I don't think Jiiks cares.

On Wed, Jan 4, 2017 at 8:59 AM, Benjamin Hoffmeyer <notifications@github.com

**Comment by ddlr** [comment]
I can also confirm it works with 0.0.13, though oddly enough, the BetterDiscord tab doesn’t appear in the User Settings.

**Comment by mkody** [comment]
Yup @ddlr, but it's broken on macOS too (at least). Discord recently pushed a background update that broke BD.

**Comment by ddlr** [comment]
@mkody Ah, good to know that I’m not the only one.

**Comment by magiruuvelvet** [comment]
There is a temporary workaround for this. The file just needs to be compied to the plugin path, it works automagically. **Don't actually enable this plugin in the BetterDiscord settings** https://gist.github.com/noodlebox/94b26b4ded85a0ded48197333aad290b

**Comment by simoniz0r** [comment]
It seems like BetterDiscord updated to v0.2.82... any chance of this PR getting updated also?  BetterDiscord won't load anymore after the update.

http://i.imgur.com/HMTkzIc.png

**Comment by MelDone** [comment]
Jup, same error :( 
screenshot from 2017-01-09 14-20-30

**Comment by mkody** [comment]
I've sent two PR #435 and #436 to fix the latest changes.

So here's what I need to do (_as of right now_) to make it run:
```sh
# Start Discord at least once and then quit before installing BetterDiscord
# Let's create a work dir
mkdir BDA
cd BDA
# Clone the upstream repo to "BetterDiscord"
git clone https://github.com/Jiiks/BetterDiscordApp.git BetterDiscord
cd BetterDiscordApp
# Now we're going to merge the PRs manually
git remote add mkody https://github.com/mkody/BetterDiscordApp.git
git remote add sarnex https://github.com/sarnex/BetterDiscordApp.git
git merge mkody/fix-bdStorage-path
git merge mkody/fix-version-config
git merge sarnex/master
# Now let's go back to our work dir and get the dependencies
cd ..
cp BetterDiscord/Installers/Node/* BetterDiscord/splice .
npm install asar wrench
# And now we can run the installer
# (you might need to use sudo)
bash install.sh
```

### Important note regarding recent changes
Plugins needs to be updated to the new storage system. Please remove them in the `~/.config/plugins` folder before launching BetterDiscord, get the latest version and add them progressively to check if they work.

**Comment by sarnex** [comment]
Sorry about that, I just rebased. Let me know if the rebased version works.

**Comment by mkody** [comment]
Yes, it still works but you need to have my two PRs merged anyways (especially #436).

Until this is if merged upstream, I'd still recommend everyone to use my instructions and merge manually.
When Jiiks and co. will push updates to BD, we can just do a `git pull` and re-run the `install.sh`.

**Comment by ddlr** [comment]
@mkody Thanks for sharing—I inserted `chmod u+x install.sh` before the last line and it worked like a charm so far. I’ll check if a BetterDiscord plugin I use still works in a moment.

**Comment by simoniz0r** [comment]
Do I still have to follow mkody's instructions to install BetterDiscord?  I've tried cloning sarnex's updated PR and reinstalling BetterDiscord, but I still get the out of date error when launching Discord.

**Comment by mkody** [comment]
Yup, #436 should fix it.

**Comment by simoniz0r** [comment]
Any chance of all of that getting merged into this PR?  I'd rather not go through all of that every time I install BetterDiscord.

**Comment by sarnex** [comment]
I merged #436 into my PR.

**Comment by simoniz0r** [comment]
Thank you so much!

**Comment by magiruuvelvet** [comment]
Still works with canary 0.0.14 :+1: 
I didn't tested the Linux PTB and Stable client. I prefer Canary :smile:

**Comment by mkody** [comment]
I guess we should change the path in `js/main.js` as well.
```diff
diff --git a/js/main.js b/js/main.js
index 4ee8bea..3f1a5d9 100644
--- a/js/main.js
+++ b/js/main.js
@@ -13,7 +13,7 @@
     let __fs = window.require("fs");
     let __process = window.require("process");
     let __platform = __process.platform;
-    let __dataPath = (__platform === 'win32' ? __process.env.APPDATA : __platform === 'darwin' ? __process.env.HOME + '/Library/Preferences' : '/var/local') + '/BetterDiscord/';
+    let __dataPath = (__platform === 'win32' ? __process.env.APPDATA : __platform === 'darwin' ? __process.env.HOME + '/Library/Preferences' : process.env.HOME + '/.config') + '/BetterDiscord/';
 
 
     let __data = {};
```

**Comment by simoniz0r** [comment]
Could you also please merge #435 into this PR?  I had to create a script that deletes `/home/user/.config/bdstorage.json` before launching BetterDiscord otherwise it complains that `/home/user/.config/BetterDiscord//bdStorage.json` doesn't exist, and gives me a javascript error that locks up the whole app.

Javascript error

**Comment by sarnex** [comment]
I merged it.

**Comment by magiruuvelvet** [comment]
And another update. Still works with 0.0.15 :+1:

**Comment by ddlr** [comment]
@magiruuvelvet I hope you’re not going to keep this up every day, considering the new update schedule. 😜

**Comment by TheDoctorsLife** [comment]
Canary it set to build on each new push to the master branch, just as an FYI.

**Comment by simoniz0r** [comment]
I don't know if you've seen this, but Ckath made a script that can install BetterDiscord to any directory... might be of some use to you.
https://github.com/Ckath/Discord-stuff/blob/master/scripts/tape.sh

**Comment by sarnex** [comment]
This is basically just doing what the BetterDiscord installer does in a
script, there's no way for it to be merged but maybe it's easier to use. I
recommend following the instructions here.

**Comment by simoniz0r** [comment]
I would say the fact that it allows any directory to be used is enough, but ¯\_(ツ)_/¯

**Comment by sarnex** [comment]
It only has to be /opt for the install, you can move Discord after.

**Comment by simoniz0r** [comment]
Sure, but that's not ideal.  That is not the default install location for the .deb package.  Also, with PTB and Stable being released on Linux, `tape.sh` makes installing BetterDiscord to those Discord versions possible without having to move them to `/opt/DiscordCanary/` first.

**Comment by Jiiks** [comment]
Maybe it's time to merge this o-o

**Comment by mkody** [comment]
Woo!
You should keep an eye on the fact that we're not using `/var/local` anymore, but `~/.config`.
This includes (at least) in `v2/lib/core.js`: `installPath` and `dataPath`, and in`js/main.js`: for `__dataPath`

**Comment by sarnex** [comment]
we did it boys, i dont mind maintaining the linux portion but please use
this PR for that

**Pull request merged**